src/Widgets/calendar/GridDay.vala

```diff
@@ -68,6 +86,55 @@ public class DateTime.Widgets.GridDay : Gtk.EventBox {
         notify["date"].connect (() => {
             label.label = date.get_day_of_month ().to_string ();
         });
+
+        event_dots = new Gee.HashMap<string, Gtk.Widget> ();
+
+        model.events_added.connect (add_event_dots);
+        model.events_removed.connect (remove_event_dots);
+    }
+
+    private void add_event_dots (E.Source source, Gee.Collection<ECal.Component> events) {
+        foreach (var component in events) {
+            if (event_dots.size >= 3) {
+                return;
+            }
+
+            unowned ICal.Component ical = component.get_icalcomponent ();
+            foreach (var dt_range in Util.event_date_ranges (ical, model.data_range)) {
```

Comment:
I wonder if this inner foreach can run many times and exceed the initial `event_dots.size >= 3` constraint? Is this something that we should be concerned with?

Reply:
Yeah moving that check seems to actually help with under-reporting the number of dots also